## 1. Importing library and functions

In [2]:
import pandas as pd
import advertools as adv
pd.options.display.max_columns = None
from devtool import g_search, yt_search


## 2. Searching queries on Google

In [5]:
query=["Sushant Singh Rajpoot", "butter chicken", "lenovo", "coronavirus", "immunity"]
df= g_search(query)

df.head(15)
# Here is a sample of the Google data showing the first ten columns. 

2021-06-15 16:33:56,706 | INFO | serp.py:727 | serp_goog | Requesting: q=Sushant Singh Rajpoot, cx=d270029c29ad7f670, key=AIzaSyDKgjA0DZa_qZO-uMJaVrV64lOFNj3-N8U
2021-06-15 16:33:57,840 | INFO | serp.py:727 | serp_goog | Requesting: q=butter chicken, cx=d270029c29ad7f670, key=AIzaSyDKgjA0DZa_qZO-uMJaVrV64lOFNj3-N8U
2021-06-15 16:33:58,596 | INFO | serp.py:727 | serp_goog | Requesting: q=lenovo, cx=d270029c29ad7f670, key=AIzaSyDKgjA0DZa_qZO-uMJaVrV64lOFNj3-N8U
2021-06-15 16:33:59,688 | INFO | serp.py:727 | serp_goog | Requesting: q=coronavirus, cx=d270029c29ad7f670, key=AIzaSyDKgjA0DZa_qZO-uMJaVrV64lOFNj3-N8U
2021-06-15 16:34:00,539 | INFO | serp.py:727 | serp_goog | Requesting: q=immunity, cx=d270029c29ad7f670, key=AIzaSyDKgjA0DZa_qZO-uMJaVrV64lOFNj3-N8U


N     NaN            NaN                  NaN   
10              NaN     NaN            NaN                  NaN   
11              NaN     NaN            NaN                  NaN   
12              NaN     NaN            NaN                  NaN   
13              NaN     NaN            NaN                  NaN   
14              NaN     NaN            NaN                  NaN   

   apple-mobile-web-app-capable og:author irsstandard-type channel prod_seid  \
0                           NaN       NaN              NaN     NaN       NaN   
1                           NaN       NaN              NaN     NaN       NaN   
2                           NaN       NaN              NaN     NaN       NaN   
3                           NaN       NaN              NaN     NaN       NaN   
4                           NaN       NaN              NaN     NaN       NaN   
5                           NaN       NaN              NaN     NaN       NaN   
6                           NaN       NaN              NaN     NaN       NaN   
7                           NaN       NaN              NaN     NaN       NaN   
8                           NaN       NaN              NaN     NaN       NaN   
9                           NaN       NaN              NaN     NaN       NaN   
10                          NaN       NaN              NaN     NaN       NaN   
11                          NaN       NaN              NaN     NaN       NaN   
12                          NaN       NaN              NaN     NaN       NaN   
13                          NaN       NaN              NaN     NaN       NaN   
14                          NaN       NaN              NaN     NaN       NaN   

   rights uniqueid audience og:image:url handheldfriendly ishistorical  \
0     NaN      NaN      NaN          NaN              NaN          NaN   
1     NaN      NaN      NaN          NaN              NaN          NaN   
2     NaN      NaN      NaN          NaN              NaN          NaN   
3     NaN      NaN      NaN          NaN              NaN          NaN   
4     NaN      NaN      NaN          NaN              NaN          NaN   
5     NaN      NaN      NaN          NaN              NaN          NaN   
6     NaN      NaN      NaN          NaN              NaN          NaN   
7     NaN      NaN      NaN          NaN              NaN          NaN   
8     NaN      NaN      NaN          NaN              NaN          NaN   
9     NaN      NaN      NaN          NaN              NaN          NaN   
10    NaN      NaN      NaN          NaN              NaN          NaN   
11    NaN      NaN      NaN          NaN              NaN          NaN   
12    NaN      NaN      NaN          NaN              NaN          NaN   
13    NaN      NaN      NaN          NaN              NaN          NaN   
14    NaN      NaN      NaN          NaN              NaN          NaN   

   abstract otid guidance mobileoptimized metatag_organization topic  \
0       NaN  NaN      NaN             NaN                  NaN   NaN   
1       NaN  NaN      NaN             NaN                  NaN   NaN   
2       NaN  NaN      NaN             NaN                  NaN   NaN   
3       NaN  NaN      NaN             NaN                  NaN   NaN   
4       NaN  NaN      NaN             NaN                  NaN   NaN   
5       NaN  NaN      NaN             NaN                  NaN   NaN   
6       NaN  NaN      NaN             NaN                  NaN   NaN   
7       NaN  NaN      NaN             NaN                  NaN   NaN   
8       NaN  NaN      NaN             NaN                  NaN   NaN   
9       NaN  NaN      NaN             NaN                  NaN   NaN   
10      NaN  NaN      NaN             NaN                  NaN   NaN   
11      NaN  NaN      NaN             NaN                  NaN   NaN   
12      NaN  NaN      NaN             NaN                  NaN   NaN   
13      NaN  NaN      NaN             NaN                  NaN   NaN   
14      NaN  NaN      NaN             NaN                  NaN   NaN 

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Columns: 304 entries, searchTerms to contentlanguage
dtypes: datetime64[ns, UTC](1), float64(1), int64(3), object(299)
memory usage: 118.9+ KB


In [7]:
print('Total number of rows:',df.shape[0], 'and columns:', df.shape[1])

Total number of rows: 50 and columns: 304


## 2.i. Creating quick summary

- `count`: The number of times the domain appeared in the dataset. Keep in mind that the same domain might appear more than once, in the same query. As you can see below YouTube appeared 80% of the queries (4 ÷ 5 unique queries).
- `avg_rank`: The average rank that the domain appeared on, in the dataset. 
- `coverage`: The number of times the domain appeared ÷ the number of unique queries sent.

In [18]:
summary = (df
           .groupby(['displayLink'], as_index=False)
           .agg({'rank': ['count', 'mean']})
           .sort_values(('rank', 'count'), ascending=False))
# creating a dataframe by grouping all link element with using agg function(count & mean) on rank column and also sorting according to the count of rank column.
           
summary.columns = ['DisplayLink', 'Count', 'Avg_Rank']
# inserting these list element as column name

summary['Avg_Rank'] = summary['Avg_Rank'].round(0)
# mean= 6.750000-----> 7.00

summary['DisplayLink'] = summary['DisplayLink'].str.replace('www.', '')
# www.fb.com----> fb.com


summary['coverage']=summary['Count'].div(df['searchTerms'].nunique()).mul(100).astype(str).add('%')
# coverage = (total count of that domain)/ (total queries we have searched) * 100 %

summary.head(10)

# yt ssr -4
# yt ssr - 6
# yt butterchkn - 9
# yt immunity - 9

# yt count--- > 4
# yt rank -----> (4+6+9+9)/4
# yt coverage -----> 4/5 *100%

<ipython-input-18-10a70ec6ad6c>:14: FutureWarning: The default value of regex will change from True to False in a future version.
  summary['DisplayLink'] = summary['DisplayLink'].str.replace('www.', '')


DisplayLink  Count  Avg_Rank coverage
37       youtube.com      4       7.0    80.0%
25        lenovo.com      3       3.0    60.0%
6   en.wikipedia.org      3       3.0    60.0%
35           who.int      2       5.0    40.0%
18      facebook.com      2       8.0    40.0%
14           cdc.gov      2       3.0    40.0%
13           bbc.com      2       6.0    40.0%
11       twitter.com      2       6.0    40.0%
28      michigan.gov      1       5.0    20.0%
24         jhsph.edu      1       8.0    20.0%

**Inference:** YouTube is clearly the undisputed leader, and very far from the first competitor. 

## 2.ii. Text Analysis

In [19]:
# Case 1: Below is the relevant word count for the Title column. After removing stopwords.

stop_words= list(adv.stopwords['english']) + ['-', '', '|', '–', '&', '.']

title_word_freq = adv.word_frequency(df['title'], rm_words=stop_words)

title_word_freq.head(15)

word  abs_freq
0        lenovo        11
1       chicken        11
2      immunity        10
3         singh        10
4        butter        10
5       sushant        10
6   coronavirus         8
7        rajput         8
8      covid-19         8
9        recipe         5
10      laptops         4
11         herd         3
12    wikipedia         3
13       indian         3
14    delicious         2

In [22]:
# Case 2: Below is the relevant word count for the Snippet column. After removing stopwords.

stop_words= list(adv.stopwords['english']) + ['-', '', '|', '–', '&', '.']

snippet_word_freq = adv.word_frequency(df['snippet'], rm_words=stop_words)

snippet_word_freq.head(15)

word  abs_freq
0       chicken        20
1      immunity        14
2        butter        12
3       sushant        10
4   coronavirus         9
5         singh         9
6      covid-19         9
7        lenovo         8
8          2020         8
9       disease         6
10       indian         6
11       rajput         6
12        known         5
13         herd         5
14        cases         4

## 3. Searching queries on YouTube

In [23]:
query=["Sushant Singh Rajpoot", "butter chicken", "lenovo", "coronavirus", "immunity"]
df2= yt_search(query)

df2.head(15)
# Here is a sample of the Youtube data showing the first ten columns. 

2021-06-15 17:24:24,246 | INFO | serp.py:1162 | serp_youtube | Requesting: key=AIzaSyDKgjA0DZa_qZO-uMJaVrV64lOFNj3-N8U, q=Sushant Singh Rajpoot
2021-06-15 17:24:26,187 | INFO | serp.py:1162 | serp_youtube | Requesting: key=AIzaSyDKgjA0DZa_qZO-uMJaVrV64lOFNj3-N8U, q=butter chicken
2021-06-15 17:24:27,077 | INFO | serp.py:1162 | serp_youtube | Requesting: key=AIzaSyDKgjA0DZa_qZO-uMJaVrV64lOFNj3-N8U, q=lenovo
2021-06-15 17:24:28,005 | INFO | serp.py:1162 | serp_youtube | Requesting: key=AIzaSyDKgjA0DZa_qZO-uMJaVrV64lOFNj3-N8U, q=coronavirus
2021-06-15 17:24:28,980 | INFO | serp.py:1162 | serp_youtube | Requesting: key=AIzaSyDKgjA0DZa_qZO-uMJaVrV64lOFNj3-N8U, q=immunity
2021-06-15 17:24:29,956 | INFO | serp.py:465 | youtube_video_details | Requesting: video details
2021-06-15 17:24:30,851 | INFO | serp.py:497 | youtube_channel_details | Requesting: channel details


://i.ytimg.com/vi/fP...                   IBC24   

                                           video.tags video.categoryId  \
0   [telugu movies, 2019 Movies, Latest Telugu Mov...               24   
1   [sushant singh rajput songs, sushant singh raj...               24   
2   [news4nation, bihar, patna, breakingnews, news...               25   
3   [Review, Movie Review By KRK, Bollywood Review...               24   
4   [lambiyaan si judaiyaan arijit singh, lambiyaa...               10   
5   [butter chicken (dish), butter chicken recipe,...               26   
6   [Butter chicken, Chicken, Butter chicken kaise...               26   
7                                                 NaN               22   
8   [easy, butter chicken, homemade butter chicken...               26   
9   [Dilliwala Butter Chicken Recipe, Cooksmart, s...               26   
10                                                NaN               20   
11                                                NaN              NaN   
12  [unbox therapy, unboxtherapy, tech, technology...               28   
13  [Restoration, Rebuild, restore, laptops, resto...               28   
14  [IBC24, Chhattisgarh News, Madhya Pradesh News...               25   

   video.liveBroadcastContent video.defaultLanguage  \
0                        none                    en   
1                        none                   NaN   
2                        none                   NaN   
3                        none                   NaN   
4                        none                   NaN   
5                        none                   NaN   
6                        none                   NaN   
7                        none                   NaN   
8                        none                   NaN   
9                        none                   NaN   
10                       none                    en   
11                        NaN                   NaN   
12                       none                    en   
13                       none                   NaN   
14                       none                   NaN   

                                      video.localized  \
0   {'title': 'Sushant Singh Rajput Funeral Video ...   
1   {'title': 'Ankita Lokhande Got Emotional For S...   
2   {'title': 'आखिरी बार Sushant singh rajpoot ने ...   
3   {'title': 'Sushant Singh Rajput Ko Shradhanjal...   
4   {'title': 'Arijit Singh : Lambiyaan Si Judaiya...   
5   {'title': 'How To Make Butter Chicken At Home ...   
6   {'title': 'Best Butter Chicken Recipe | ढाबा स...   
7   {'title': 'Butter Chicken Recipe | How to make...   
8   {'title': 'Restaurant Style Butter Chicken | ब...   
9   {'title': 'Dilliwala Butter Chicken Recipe | C...   
10  {'title': 'Am I Levinho? | PUBG MOBILE', 'desc...   
11                                                NaN   
12  {'title': 'Lenovo Legion Phone Duel 2 Unboxing...   
13  {'title': 'Restoration a destroyed 20-year-old...   
14  {'title': 'Coronavirus India Update : 24 घंटे ...   

   video.defaultAudioLanguage video.viewCount video.likeCount  \
0                          te          783086           18466   
1                          hi         1723143           36117   
2                          hi         5442288           35903   
3                         NaN          111723           22097   
4                         NaN       159114064         1475735   
5                          en        31239786          417217   
6                       en-GB         2246791             NaN   
7                       en-GB         1693531           52736   
8                          hi         7188099          118437   
9                         NaN         8967745          138522   
10                         en          336291           51081   
11                        NaN             NaN             NaN   
12                         en         2513997           30288   
13                        NaN         9125666          1516

In [24]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 85 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   q                              24 non-null     object             
 1   queryTime                      24 non-null     datetime64[ns, UTC]
 2   rank                           24 non-null     int64              
 3   title                          24 non-null     object             
 4   description                    24 non-null     object             
 5   publishedAt                    24 non-null     object             
 6   channelTitle                   24 non-null     object             
 7   totalResults                   24 non-null     int64              
 8   kind                           24 non-null     object             
 9   channelId                      24 non-null     object             
 10  default.height              

In [25]:
print('Total number of rows:',df2.shape[0], 'and columns:', df2.shape[1])

Total number of rows: 24 and columns: 85


**Inference : ** Note the difference in the number of columns here. 85 compared to 304 on Google search. 

## 3.i. Creating quick summary

- `count`: The number of times the domain appeared in the dataset. Keep in mind that the same domain might appear more than once, in the same query.
- `avg_rank`: The average rank that the domain appeared on, in the dataset. 
- `coverage`: The number of times the domain appeared ÷ the number of unique queries sent.

In [27]:
summary2 = (df2
           .groupby(['channelTitle'], as_index=False)
           .agg({'rank': ['count', 'mean']})
           .sort_values(('rank', 'count'), ascending=False))
# creating a dataframe by grouping all link element with using agg function(count & mean) on rank column and also sorting according to the count of rank column.
           
summary2.columns = ['Channel_Name', 'Count', 'Avg_Rank']
# inserting these list element as column name


summary2['Avg_Rank'] = summary2['Avg_Rank'].round(0)
# mean= 6.750000-----> 7.00


summary2['Coverage']=summary2['Count'].div(df2['q'].nunique()).mul(100).astype(str).add('%')
# coverage = (total count of that domain)/ (total queries we have searched) * 100 %
# df2['q'].nunique() === len(query) === 5

summary2.head(15)

# TSeries ssr -4
# TSeries ssr - 6
# TSeries butterchkn - 9
# TSeries immunity - 9

# TSeries count--- > 4
# TSeries rank -----> (4+6+9+9)/4
# TSeries coverage -----> 4/5 *100%

Channel_Name  Count  Avg_Rank Coverage
0         9 News Australia      2       4.0    40.0%
11           MEDSimplified      2       3.0    40.0%
16         Shomu's Biology      2       2.0    40.0%
19          Viralbollywood      1       2.0    20.0%
18           Unbox Therapy      1       4.0    20.0%
17                T-Series      1       5.0    20.0%
15  Sanjeev Kapoor Khazana      1       5.0    20.0%
14       Restoration T - D      1       5.0    20.0%
13             News4Nation      1       3.0    20.0%
12       Movie Time Cinema      1       1.0    20.0%
10                 Levinho      1       1.0    20.0%
1        Chef Ranveer Brar      1       2.0    20.0%
9                   Lenovo      1       2.0    20.0%
8              Kunal Kapur      1       4.0    20.0%
7      Kamaal R Khan - KRK      1       4.0    20.0%

**Inference : ** Here the picture is completely different from Google search. YouTube returns five results per query, so this also has an effect on how data are distributed.

## 3.ii. Text Analysis

Things get more interesting with weighted frequency.  
Here is the word frequency for YouTube's DataFrame, counting words in the title column, with videos' number of views.

In [30]:
# Case 1: Below is the relevant word count in Youtube video's title. After removing stopwords.

stop_words = list(adv.stopwords['english']) + ['-', '', '|', '–', '&']

ytb_title_word_freq = adv.word_frequency(text_list=df2['title'], rm_words=stop_words)

ytb_title_word_freq.head(15)

word  abs_freq
0       chicken         9
1        butter         7
2      immunity         6
3       sushant         5
4   coronavirus         5
5         singh         5
6        rajput         4
7        lenovo         4
8        recipe         4
9        innate         3
10         chef         3
11       system         3
12         home         3
13      records         2
14     victoria         2

In [42]:
# Case 2: Below is the relevant word count in Youtube video's description. After removing stopwords.

stop_words = list(adv.stopwords['english']) + ['-', '', '|', '–', '&']

ytb_desc_word_freq = adv.word_frequency(text_list=df2['description'],rm_words=stop_words)
# df2.'description=['ssr died in 2020', 'ssr died by suicide'......]

ytb_desc_word_freq.head(15)


word  abs_freq
0    chicken        10
1       news         7
2   immunity         7
3     butter         7
4          •         5
5     latest         4
6     recipe         4
7      world         4
8     lenovo         4
9      video         4
10    system         3
11     singh         3
12    follow         3
13   lecture         3
14     human         3

In [45]:
# Case 3: Below is the relevant video tags count in Youtube video. After removing stopwords.

stop_words = list(adv.stopwords['english']) + ['-', '', '|', '–', '&']

list_of_tags=[]
for i in df2['video.tags'].fillna(""):
    list_of_tags+=i

ytb_tags_word_freq = adv.word_frequency(text_list=list_of_tags , rm_words=stop_words)

ytb_tags_word_freq.head(15)

word  abs_freq
0       chicken        57
1        butter        48
2         singh        31
3          news        30
4       sushant        27
5        system        25
6      immunity        23
7        rajput        23
8        recipe        23
9        immune        18
10   complement        14
11  coronavirus        13
12           si        10
13     covid-19        10
14        covid         8

In [38]:
# Case 4: Below is the relevant Hashtags count in Youtube video's description. After removing stopwords.

stop_words = list(adv.stopwords['english']) + ['-', '', '|', '–', '&']

ytb_hashtags_word_freqadv= adv.word_frequency(df2['video.description'].fillna(' '), regex=adv.regex.HASHTAG_RAW, rm_words=stop_words)

ytb_hashtags_word_freqadv.head(15)

word  abs_freq
0        #butterchicken         4
1             #9newsaus         2
2        #chickenrecipe         2
3    #ninenewsaustralia         2
4         #breakingnews         2
5          #coronavirus         2
6                  #krk         1
7        #coronainindia         1
8           #pubgmobile         1
9              #levinho         1
10  #coronavirusinindia         1
11       #covid19update         1
12        #coronaupdate         1
13              #corona         1
14       #deathsinindia         1

**Inference : **This was a quick overview of doing some SEO research on Google and YouTube, and how you can compare, and extract certain information.  
There is more that can be done but these are some of the main features that I think are important.  